src/test/codewhisperer/util/authUtil.test.ts

```diff
@@ -0,0 +1,180 @@
+    )
+    provider.invalidate.callsFake(async () => (token = undefined))
+
+    return provider
+}
+
+function getTestTokenProvider(...[profile]: ConstructorParameters<typeof SsoAccessTokenProvider>) {
+    const key = getSsoProfileKey(profile)
+    const cachedProvider = tokenProviders.get(key)
+    if (cachedProvider !== undefined) {
+        return cachedProvider
+    }
+
+    const provider = createTestTokenProvider()
+    tokenProviders.set(key, provider)
+
+    return provider
+}
+
+let auth: Auth
+let store: ProfileStore
+let builderIdConn: SsoConnection | undefined
+let ssoConn: SsoConnection
+let entSsoConn: SsoConnection
+let mockConnListNoBuilder: Connection[]
+let mockConnListNoEntSso: Connection[]
+const ssoProfile = createSsoProfile()
+const awsBuilderIdProfile = createAwsBuilderIdProfile({ scopes: codewhispererScopes })
+const enterpriseSsoProfile = createEntSsoProfile({ scopes: codewhispererScopes })
+
+async function createMockConnections(){
+    store = new ProfileStore(new FakeMemento())
+    auth = new Auth(store, getTestTokenProvider, new CredentialsProviderManager())
+    builderIdConn = await auth.createConnection(awsBuilderIdProfile)
+    ssoConn = await auth.createConnection(ssoProfile)
+    entSsoConn = await auth.createConnection(enterpriseSsoProfile)
+}
+
+describe('AuthUtil', async function () {
+   before(async function () {
+       await createMockConnections()
+   })
+
+    beforeEach(async function () {
+        sinon.stub(Commands, 'register') 
+        store = new ProfileStore(new FakeMemento())
+        auth = new Auth(store, getTestTokenProvider, new CredentialsProviderManager())
+        
+        sinon.replace(SsoClient, 'create', () => {
+            const s = stub(SsoClient)
+            s.logout.resolves()
+
+            return s
+        })
+              
+    })
+
+    afterEach(async function () {
+        sinon.restore()
+    })
+
+    //TODO
+    //factory function to create connectionlists dynamically
+    //stub secondaryAuth useNewConnection? vs auth instance use connection
+    //remove auth and profile from beforeEach and see if it breaks
+    //todo: figure out what sinon.replace in beforeEach does
+    //add start url to params of create SSO profile
+    //figure out where to declare all the vars
+    //TODO add more intermittent asserts
+    it('if there is no valid AwsBuilderID conn, it will create one and use it', async function () {
+        mockConnListNoBuilder = [ssoConn]
+        sinon.stub(Auth.instance, 'listConnections').resolves(mockConnListNoBuilder)
+        sinon.stub(Auth.instance, 'createConnection').resolves(builderIdConn)
+        const authSpy = sinon.stub(Auth.instance, 'useConnection')
+
+        getTestWindow().onDidShowQuickPick(async picker => {
+            await picker.untilReady()
+            picker.acceptItem(picker.items[1])
+        })
+
+        await AuthUtil.instance.connectToAwsBuilderId()
```

Comment:
You shouldn't need to stub anything
```suggestion
        const authUtil = new AuthUtil(auth)
        await authUtil.connectToAwsBuilderId()
```